libmcount/dynamic.c

```diff
@@ -327,12 +354,15 @@ static void prepare_dynamic_update(struct uftrace_sym_info *sinfo, bool needs_mo
 	};
 	int hash_size = sinfo->exec_map->mod->symtab.nr_sym * 3 / 4;
 
-	if (needs_modules)
-		hash_size *= 2;
-
-	code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	if (code_hmap && !needs_modules) /* already prepared */
+		return;
+	if (!code_hmap) { /* nothing loaded */
+		if (needs_modules)
+			hash_size *= 2;
+		code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	}
 
-	dl_iterate_phdr(find_dynamic_module, &fmd);
+	modules_loaded = !dl_iterate_phdr(find_dynamic_module, &fmd);
```

Comment:
I'm not sure if this is correct.  IIRC the return value is from the (last) callback.  Maybe you can set it only if it needs modules.  How about this?

```c
if (code_hmap) {
    /* main executable should be loaded already */
    if (!needs_modules || modules_loaded)
        return;

    /* maybe increase the hash map size */
}
else {
    if (needs_modules)
        hash_size *= 2;
    code_hmap = hashmap_create(...);
}

dl_iterate_phdr(...);

main_loaded = true;
if (needs_modules)
    modules_loaded = true;
```